### DISCUSSION TITLE
How to use FusionCache with prefetching?

### DISCUSSION THREAD

**Discussion by hvgNET** [ZiggyCreatures/FusionCache#69] [Q&A] [answered]
(This is my first time using GitHub, so if I'm doing something wrong, please don't hesitate to tell me)

Hello,

I am considering using FusionCache for a project and I am currently testing it out.  
I have a scenario that I don't know how to solve with my current knowledge about FusionCache, hence this question.  

The scenario is that I want to prefill the cache with data, but then let the GetOrSet factory stuff handle cache misses, timeouts, cache refreshes and all that good stuff after the cache have been filled.
I want to prefill the cache from a database and the important detail about this scenario is that i only want to call the database **once**, when prefilling.

Example:

I have 10.000.000 key value pairs in a db, and i know the ids upfront.
I have a method that takes a single id called GetValue(int id), which accesses the db.

I could prefill the cache by doing GetOrSet({id}, _ => GetValue(id)) for each of the ids.  
But to my understanding this would go to the db 10M times, which is the reason I want to go to the db once.

So conceptually I would like to do something like 
* Call a method called GetAllValues()
* Use Set to set them individually for each of them
* Call GetOrSet({id}, _ => GetValue(id)) for each of them

But that just seems... a bit hacky to me as i neither want to get nor set the value, i just want to set the factory up

I hope it makes sense, else I am happy to provide further information or input :)

**Comment by jodydonetti** [comment, marked as answer]
Hi @hvgNET and thanks for considering FusionCache.

Sorry for the delay in answering but covid finally got me.

In general your idea seems the right one to me:
- load all 10.000.000 items from the db in one go
- save them all in the cache individually with a `Set` call for each of them (with a reasonable cache duration)

Then, when you need one of them during the lifetime of your application, call `GetOrSet({id}, _ => GetValue(id))` so that if the value is still in the cache you'll get it back immediately and, if not in the cache anymore, it will be loaded from the db and saved in the cache.

There's this passage that I don't understand though:

> But that just seems... a bit hacky to me as i neither want to get nor set the value, i just want to set the factory up

Why do you say it feels hacky? Just to be clear, you don't need to call `GetOrSet` for each item when you are prefetching the entire thing at the beginning, you only need to call `GetOrSet` later on when you need some of those values and, in that moment, the `GetOrSet` means _"**get** me X from the cache and, if it is not there, load it and **set** it"_.

Maybe you are thinking about calling `GetOrSet` for each key in the initial prefetching phase? If that is the case, that is not necessary.

So, to recap:

- INITIAL PREFETCHING: load all items from db -> for each item call `Set` with a reasonable duration
- DURING YOUR NORMAL APP USAGE: call `GetOrSet` so that if the value is expired from the cache it will be loaded again

Doing it this way will allow you to have the cache already filled at start + automatically keep some of those items in the cache, based on actual usage.

Let me know if I got this right and if I can help you more.

Hope this helps.

**Reply by hvgNET** [reply]
Thanks for your reply and I hope you are feeling better :)

You understood my "But that just seems... a bit hacky" correctly.  
The reason why I wrote it seems a bit hacky, is because in my understanding I thought that e.g. the FailSafe options and AllowTimedOutFactoryBackgroundCompletion was only in effect if it had been used together with GetOrSet to set up the cache entry.

But from what I understand from your answer, these options will be considered as long as i pass them with the GetOrSet call if it is a cache miss.  
It makes a lot more sense than my previous understanding, I think I just got confused somewhere down the road.  
So thank you very much for clarifying :)

**Reply by jodydonetti** [reply]
Happy that helped 🙂

Since we are on these details, I'll add one more just to complete the picture.

To function properly, fail-safe must be enabled BOTH:
- when saving some data in the cache (so that FusionCache knows to keep it around for some extra time after expiration)
- AND when requesting the same data later on, so that if it's the case it can be activated

Typically this is normal behavior, because you either want fail-safe for a piece of data (or for every piece of data) or not, but I wanted to be extra sure you got the full picture. So if you SET something with fail-safe disabled you cannot expect it to work later on with a GET, becuase the data has not been kept around (this is done because I don't want to keep around data if that is not desiderd, since it would be a waste of resources).

I would think about it this way:
- when saving something in the cache (SET) if you enable fail-safe you are saying _"keep it around some extra time, so it may be used later"_ (just in case, we don't know yet)
- when getting something from the cache (GET) if you enable fail-safe you are saying _"if something goes wrong and there's data kept around, use that (instead of throwing an error)"_
- of course a `GetOrSet` is a GET and potentially a SET, so the 2 things may apply together.

Let me know if you need something more.

ps: and yes, I'm feeling better, thanks!

**Comment by jodydonetti** [comment]
**ps:** if you plan on also using a 2nd layer (distributed cache) keep in mind that the initial prefetching may not be necessary, since the data may already be in the distributed cache because of previous runs.
